sanic/testing.py

```diff
@@ -10,17 +11,21 @@
 
 class SanicTestClient:
     def __init__(self, app, port=PORT):
+        """Use port=None to bind to a random port"""
         self.app = app
         self.port = port
 
-    async def _local_request(self, method, uri, cookies=None, *args, **kwargs):
+    async def _local_request(
+        self, method, uri, sock=None, cookies=None, *args, **kwargs
```

Comment:
My main issue here is that this function signature now has duplicate information, it seems like if we have all the information we need in the caller to infer the url then it should be the callers responsibility for doing so.The easiest way to do that would be `self.port = sock.getsockname()[1]` after 92, lazily setting the port _should_ be fine but it may be worth thinking about. Another option though is we have a short circuit on 23 for uri's that look like urls, we could similarly do `uri = "http://{sock[0]}:{sock[1]}{uri}".format(sock=sock.getaddrname(), uri=uri)` at after 92 as well.

Reply:
I like the idea, but setting `self.port` causes multiple requests to bind the same port, which may fail if the OS assigned the random port to something else between requests. I think it would be better to set `self.sock` to `None` in `__init__` and set and unset it from `_collect_request()` below.

Reply:
> we could similarly do `uri = "http://{sock[0]}:{sock[1]}{uri}".format(sock=sock.getaddrname(), uri=uri)` at after 92 as well.

I ended up going with that, more or less